rust/src/dcerpc/dcerpc.rs

```diff
@@ -0,0 +1,2600 @@
+        }
+        self.bytes_consumed += retval as u16;
+
+        // If the query has been completed, clean the buffer and reset the direction
+        if self.query_completed == true {
+            self.clean_buffer(direction);
+            self.reset_direction(direction);
+        }
+        self.prev_dir = direction;
+        return AppLayerResult::ok();
+    }
+}
+
+#[derive(Debug)]
+pub struct DCERPCUDPState {
+    pub header: Option<DCERPCHdrUdp>,
+    pub request: Option<DCERPCRequest>,
+    pub response: Option<DCERPCResponse>,
+    pub fraglenleft: u16,
+    pub uuid_entry: Option<DCERPCUuidEntry>,
+    pub uuid_list: Vec<DCERPCUuidEntry>,
+    pub de_state: Option<*mut core::DetectEngineState>,
+}
+
+impl DCERPCUDPState {
+    pub fn new() -> DCERPCUDPState {
+        return DCERPCUDPState {
+            header: None,
+            request: None,
+            response: None,
+            fraglenleft: 0,
+            uuid_entry: None,
+            uuid_list: Vec::new(),
+            de_state: None,
+        };
+    }
+
+    fn new_request(&mut self) {
+        let request = DCERPCRequest::new();
+        self.request = Some(request);
+    }
+
+    fn new_response(&mut self) {
+        let response = DCERPCResponse::new();
+        self.response = Some(response);
+    }
+
+    fn create_new_query(&mut self, pkt_type: u8) {
+        match pkt_type {
+            DCERPC_TYPE_REQUEST => {
+                self.new_request();
+            }
+            DCERPC_TYPE_RESPONSE => {
+                self.new_response();
+            }
+            _ => {
+                SCLogDebug!("Unrecognized packet type");
+            }
+        }
+    }
+
+    fn get_hdr_pkt_type(&self) -> Option<u8> {
+        if let Some(ref hdr) = &self.header {
+            return Some(hdr.pkt_type);
+        }
+        // Shouldn't happen
+        None
+    }
+
+    fn get_hdr_rpc_vers(&self) -> Option<u8> {
+        if let Some(ref hdr) = &self.header {
+            return Some(hdr.rpc_vers);
+        }
+        // Shouldn't happen
+        None
+    }
+
+    fn get_hdr_flags1(&self) -> Option<u8> {
+        if let Some(ref hdr) = &self.header {
+            return Some(hdr.flags1);
+        }
+        // Shouldn't happen
+        None
+    }
+
+    pub fn get_hdr_fraglen(&self) -> Option<u16> {
+        if let Some(ref hdr) = &self.header {
+            return Some(hdr.fraglen);
+        }
+        // Shouldn't happen
+        None
+    }
+
+    pub fn handle_fragment_data(&mut self, input: &[u8], input_len: u16) -> u16 {
+        let mut retval: u16 = 0;
+        let hdrflags1 = match self.get_hdr_flags1() {
+            Some(x) => x,
+            None => 0, // TODO check if this is OK
+        };
+        let rpc_vers = match self.get_hdr_rpc_vers() {
+            Some(x) => x,
+            None => 0,
+        };
+        let fraglenleft = self.fraglenleft;
+
+        // Update the stub params based on the packaet type
+        match self.get_hdr_pkt_type() {
```

Comment:
Wonder if this double `match` could be collapsed into `match self.get._hdr_ptr_type().unwrap_or(0)` so that the result would be `x` instead of `Option(x)`?